physicsnemo/distributed/shard_tensor.py

```diff
@@ -305,132 +339,123 @@ def __repr__(self) -> str:
+            dtensor._spec.mesh,
+            dtensor._spec.placements,
+            sharding_shapes="chunk",
+            global_shape=dtensor.shape,
+        )
 
-            cls._enable_shard_patches = True
+        return ShardTensor.__new__(
+            cls,
+            local_tensor=dtensor._local_tensor,
+            spec=spec,
+            requires_grad=dtensor.requires_grad,
+        )
 
-            return ShardTensor.__new__(
-                cls,
-                local_tensor=dtensor._local_tensor,
-                spec=spec,
-                requires_grad=dtensor.requires_grad,
-            )
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs={}):
+        with annotate(f"__torch_function___{func.__name__}"):
+            # Check for overrides:
+            if func in cls._function_registry and cls._enable_shard_patches:
+                res = cls._function_registry[func](func, types, args, kwargs)
+                return res
+            # Fall back to the default behavior:
+            return super().__torch_function__(func, types, args, kwargs)
 
     @classmethod
     @torch._disable_dynamo
+    @profile
     def __torch_dispatch__(
         cls,
         func: torch._ops.OpOverload,
         types: Tuple[type, ...],
         args: Tuple[object, ...] = (),
         kwargs: Optional[Dict[str, object]] = None,
     ) -> Union["ShardTensor", Iterable["ShardTensor"], object]:
-        # Leverage DTensor Dispatch as much as possible, but, enable
-        # the ability to operate on this output in the future:
-
-        if func in cls._function_registry:
-            return cls._function_registry[func](*args, **kwargs)
-
-        dispatch_res = DTensor._op_dispatcher.dispatch(func, args, kwargs or {})
-
-        # dispatch_res = ShardTensor._op_dispatcher.dispatch(func, args, kwargs or {})
-
-        # Return a shard tensor instead of a dtensor.
-        # ShardTensor inherits from DTensor and can lazy-init from for efficiency
-        if isinstance(dispatch_res, DTensor):
-            return ShardTensor.from_dtensor(dispatch_res, force_sharding_inference=True)
-
-        if isinstance(dispatch_res, Iterable):
-            return type(dispatch_res)(
-                ShardTensor.from_dtensor(d, force_sharding_inference=True)
-                if isinstance(d, DTensor)
-                else d
-                for d in dispatch_res
-            )
+        with annotate(f"__torch_dispatch___{func.__name__}"):
+            # Leverage DTensor Dispatch as much as possible, but, enable
+            # the ability to operate on this output in the future:
+            if func in cls._dispatch_registry:
+                res = cls._dispatch_registry[func](*args, **kwargs)
+                return res
+
+            # We assume that if we reach this point, the operator has not been
+            # intercepted by a wrapper or in the registry.  So the DTensor
+            # default behavior is likely to be correct.
```

Comment:
Would this be worth a `warnings.warn()`? How often do you expect to have ShardTensors use functions that are not in the registry, and where they cannot fall back on the DTensor implementation and still sort-of-work? As they say: being loudly wrong is ok, being silently wrong is not. This particular case is a judgment call that you're in a better position to make than I am, but just food for thought

Reply:
In this case, we definitely don't want to emit a warning.  In fact, I rely on DTensor for so many of these operations that to put a warning here would just be ignored, since it would warn with just about every single operation.

As an example: `a + b` when both are shard tensors -> fall back to DTensor.
Nearly every single-tensor function or operation (`2*a`, `torch.nn.functional.relu(a)`, etc) -> fall back to DTensor.

The reason I have overloads, in general, is that some assumptions in DTensor break things (chunk sharding sizes are mandatory) while other behavior is incorrect for domain parallelism (like in the reductions, making sure grads go the right way and the mean is weighted properly).

tl;dr falling back to DTensor is a key design feature, actually :)